File: src/main/scala/za/co/absa/hyperdrive/trigger/api/rest/services/JobTemplateService.scala
/*
 * Copyright 2018 ABSA Group Limited
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package za.co.absa.hyperdrive.trigger.api.rest.services

import org.springframework.stereotype.Service
import za.co.absa.hyperdrive.trigger.models.search.{TableSearchRequest, TableSearchResponse}
import za.co.absa.hyperdrive.trigger.models.{DagDefinitionJoined, JobTemplate, ResolvedJobDefinition, Workflow}
import za.co.absa.hyperdrive.trigger.persistance.JobTemplateRepository

import scala.concurrent.{ExecutionContext, Future}

trait JobTemplateService {
  val jobTemplateRepository: JobTemplateRepository
  val jobTemplateValidationService: JobTemplateValidationService

  def getJobTemplate(id: Long)(implicit ec: ExecutionContext): Future[JobTemplate]
  def resolveJobTemplate(dagDefinition: DagDefinitionJoined)(implicit ec: ExecutionContext): Future[Seq[ResolvedJobDefinition]]
  def getJobTemplates()(implicit ec: ExecutionContext): Future[Seq[JobTemplate]]
  def getJobTemplatesByIds(ids: Seq[Long])(implicit ec: ExecutionContext): Future[Seq[JobTemplate]]
  def getJobTemplateIdsByNames(names: Seq[String])(implicit ec: ExecutionContext): Future[Map[String, Long]]
  def searchJobTemplates(searchRequest: TableSearchRequest)(implicit ec: ExecutionContext): Future[TableSearchResponse[JobTemplate]]
  def createJobTemplate(jobTemplate: JobTemplate)(implicit ec: ExecutionContext): Future[JobTemplate]
  def updateJobTemplate(jobTemplate: JobTemplate)(implicit ec: ExecutionContext): Future[JobTemplate]
  def deleteJobTemplate(id: Long)(implicit ec: ExecutionContext): Future[Boolean]
  def getJobTemplateUsage(id: Long): Future[Seq[Workflow]]
}

@Service
class JobTemplateServiceImpl(override val jobTemplateRepository: JobTemplateRepository, jobTemplateResolutionService: JobTemplateResolutionService, override val jobTemplateValidationService: JobTemplateValidationService) extends JobTemplateService with UserDetailsService {
  override def getJobTemplate(id: Long)(implicit ec: ExecutionContext): Future[JobTemplate] = {
    jobTemplateRepository.getJobTemplate(id)
  }

  override def resolveJobTemplate(dagDefinitionJoined: DagDefinitionJoined)(implicit ec: ExecutionContext): Future[Seq[ResolvedJobDefinition]] = {
    val jobTemplateIds = dagDefinitionJoined.jobDefinitions.flatMap(_.jobTemplateId)
    jobTemplateRepository.getJobTemplatesByIds(jobTemplateIds).map(
      jobTemplates => jobTemplateResolutionService.resolveDagDefinitionJoined(dagDefinitionJoined, jobTemplates)
    )
  }

  override def getJobTemplates()(implicit ec: ExecutionContext): Future[Seq[JobTemplate]] =
    jobTemplateRepository.getJobTemplates()

  override def getJobTemplatesByIds(ids: Seq[Long])(implicit ec: ExecutionContext): Future[Seq[JobTemplate]] =
    jobTemplateRepository.getJobTemplatesByIds(ids)

  override def getJobTemplateIdsByNames(names: Seq[String])(implicit ec: ExecutionContext): Future[Map[String, Long]] =
    jobTemplateRepository.getJobTemplateIdsByNames(names)

  override def searchJobTemplates(searchRequest: TableSearchRequest)(implicit ec: ExecutionContext): Future[TableSearchResponse[JobTemplate]] = {
    jobTemplateRepository.searchJobTemplates(searchRequest)
  }

  override def createJobTemplate(jobTemplate: JobTemplate)(implicit ec: ExecutionContext): Future[JobTemplate] = {
    val userName = getUserName.apply()
    for {
      _ <- jobTemplateValidationService.validate(jobTemplate)
      jobTemplateId <- jobTemplateRepository.insertJobTemplate(jobTemplate, userName)
    } yield {
      jobTemplate.copy(id = jobTemplateId)
    }
  }

  override def updateJobTemplate(jobTemplate: JobTemplate)(implicit ec: ExecutionContext): Future[JobTemplate] = {
    val userName = getUserName.apply()
    for {
      _ <- jobTemplateValidationService.validate(jobTemplate)
      _ <- jobTemplateRepository.updateJobTemplate(jobTemplate, userName)
    } yield {
      jobTemplate
    }
  }

  override def deleteJobTemplate(id: Long)(implicit ec: ExecutionContext): Future[Boolean] = {
    val userName = getUserName.apply()
    jobTemplateRepository.deleteJobTemplate(id, userName).map(_ => true)
  }

  override def getJobTemplateUsage(id: Long): Future[Seq[Workflow]] =
    jobTemplateRepository.getJobTemplateUsage(id)
}

File: ui/src/app/services/job-template/job-template.service.ts
/*
 * Copyright 2018 ABSA Group Limited
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import { Injectable } from '@angular/core';
import { HttpClient, HttpErrorResponse, HttpParams } from '@angular/common/http';
import { api } from '../../constants/api.constants';
import { catchError, map } from 'rxjs/operators';
import { Observable, throwError } from 'rxjs';
import { JobTemplateModel } from '../../models/jobTemplate.model';
import { TableSearchRequestModel } from '../../models/search/tableSearchRequest.model';
import { TableSearchResponseModel } from '../../models/search/tableSearchResponse.model';
import { HistoryModel, HistoryPairModel } from '../../models/historyModel';
import { JobTemplateHistoryModel } from '../../models/jobTemplateHistoryModel';
import { WorkflowModel } from '../../models/workflow.model';

@Injectable({
  providedIn: 'root',
})
export class JobTemplateService {
  constructor(private httpClient: HttpClient) {}

  searchJobTemplates(searchRequestModel: TableSearchRequestModel): Observable<TableSearchResponseModel<JobTemplateModel>> {
    return this.httpClient
      .post<TableSearchResponseModel<JobTemplateModel>>(api.SEARCH_JOB_TEMPLATES, searchRequestModel, {
        observe: 'response',
      })
      .pipe(
        map((_) => {
          return _.body;
        }),
      );
  }

  getJobTemplate(id: number): Observable<JobTemplateModel> {
    const params = new HttpParams().set('id', id.toString());

    return this.httpClient
      .get<JobTemplateModel>(api.GET_JOB_TEMPLATE, { params: params, observe: 'response' })
      .pipe(map((response) => response.body));
  }

  createJobTemplate(jobTemplate: JobTemplateModel): Observable<JobTemplateModel> {
    return this.httpClient
      .put<JobTemplateModel>(api.CREATE_JOB_TEMPLATE, jobTemplate, { observe: 'response' })
      .pipe(
        map((_) => {
          return _.body;
        }),
        catchError((errorResponse: HttpErrorResponse) => {
          return throwError(errorResponse.error);
        }),
      );
  }

  updateJobTemplate(jobTemplate: JobTemplateModel): Observable<JobTemplateModel> {
    return this.httpClient
      .post<JobTemplateModel>(api.UPDATE_JOB_TEMPLATE, jobTemplate, { observe: 'response' })
      .pipe(
        map((_) => {
          return _.body;
        }),
        catchError((errorResponse: HttpErrorResponse) => {
          return throwError(errorResponse.error);
        }),
      );
  }

  deleteJobTemplate(id: number): Observable<boolean> {
    const params = new HttpParams().set('id', id.toString());

    return this.httpClient
      .delete<boolean>(api.DELETE_JOB_TEMPLATE, { params: params, observe: 'response' })
      .pipe(map((_) => _.body));
  }

  getHistoryForJobTemplate(jobTemplateId: number): Observable<HistoryModel[]> {
    const params = new HttpParams().set('jobTemplateId', jobTemplateId.toString());

    return this.httpClient
      .get<HistoryModel[]>(api.GET_HISTORY_FOR_JOB_TEMPLATE, { params: params, observe: 'response' })
      .pipe(map((_) => _.body));
  }

  getJobTemplatesFromHistory(leftHistoryId: number, rightHistoryId: number): Observable<HistoryPairModel<JobTemplateHistoryModel>> {
    const params = new HttpParams()
      .set('leftJobTemplateHistoryId', leftHistoryId.toString())
      .set('rightJobTemplateHistoryId', rightHistoryId.toString());

    return this.httpClient
      .get<HistoryPairModel<JobTemplateHistoryModel>>(api.GET_JOB_TEMPLATES_FROM_HISTORY, { params: params, observe: 'response' })
      .pipe(map((_) => _.body));
  }

  getJobTemplateUsage(id: number): Observable<WorkflowModel[]> {
    return this.httpClient
      .get<WorkflowModel[]>(api.GET_JOB_TEMPLATE_USAGE.replace('{id}', id.toString()), { observe: 'response' })
      .pipe(map((response) => response.body));
  }
}
